File: sunqi/addBinary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        string ans;
        int carry = 0;
        int sum = 0;
        int i = a.size() - 1, j = b.size() - 1;
        while (!(i < 0 && j < 0)) {
            int ai = i >= 0 ? a[i] - '0' : 0;
            int bj = j >= 0 ? b[j] - '0' : 0;
            sum = ai + bj + carry;
            ans.insert(ans.begin(), sum % 2 + '0');
            carry = sum / 2;
            --i, --j;
        }
        if (carry == 1) {
            ans.insert(ans.begin(), '1');
        }
        return ans;
    }
};

File: wuzhentao/addBinary.swift
class Solution {
    func addBinary(_ a: String, _ b: String) -> String {
        let a1 = Array(a.utf8CString.dropLast().reversed()).map( { Int($0 - Int8(48)) })
        let len_a = a1.count
        let b1 = Array(b.utf8CString.dropLast().reversed()).map( { Int($0 - Int8(48)) })
        let len_b = b1.count
        var cal: [Int] = []
        var carry = 0
        let count = max(len_b, len_a)
        
        for i in 0..<count {
            if i < len_a {
                carry += a1[i]
            }
            
            if i < len_b {
                carry += b1[i]
            }
            cal.append(carry % 2)
            carry /= 2
        }
        
        if carry != 0 {
            cal.append(1)
        }
        
        var result:String=""
        cal.reversed().forEach { (s) in
            result.append(String(s))
        }
        return result
    }
}
